aten/src/ATen/native/ReduceOps.cpp

```diff
@@ -287,7 +284,7 @@ Tensor cumprod_backward(const Tensor& grad, const Tensor& input, int64_t dim, co
     return grad;
   }
   dim = at::maybe_wrap_dim(dim, input.dim());
-  const int64_t dim_size = input.size(dim);
+  const int64_t dim_size = input.sizes()[dim];
```

Comment:
Why change that?

Reply:
It's marginally faster when `dim >= 0`, and we are calling `wrap_dim` in the line before.

Reply:
Why is it faster? It feels like it creates a temporary vector that you then index.
While the first one just get that number directly?

Reply:
I think I read that `sizes()[dim]` was more performant in a comment in a PR or in the code, and I went on and looked into the code and convinced myself that that was the case. But now that I'm looking into it again, and the code does a `maybe_wrap_dim`, and then it calls into `size_at_unchecked`, which does
https://github.com/pytorch/pytorch/blob/bdb964f89f6ae2650ab73219cb28c49774b68547/c10/core/impl/SizesAndStrides.h#L210-L212
For some reason, I got confused and thought that this was also creating a small buffer, but it's just doing pointer arithmetic.

Sorry for that, reverting.

Reply:
I may be wrong this this one!
cc @ngimel  might know better?

Reply:
Actually, now I remember what was the reason. `sizes()` does not create a temporary vector, but an `ArrayRef` (note that we are not calling `.vec()`). As such, That is, in the worst case scenario, the copy of a pointer and a size:
https://github.com/pytorch/pytorch/blob/bdb964f89f6ae2650ab73219cb28c49774b68547/c10/util/ArrayRef.h#L76-L78
and then the indexing.

The indirection is done on an xvalue, and given that we have a pointer to const T, it should not be too difficult to be optimised away by the compiler.

On the other hand, it seems difficult to optimise away a number of if-elses that are in the `size(idx)`, in particular that of the `maybe_wrap_dim`.

All in all, the difference between each should be negligible, and perhaps readability would count more in this case, as this is not in a hotpath. That being said, it'd be nice to know what's the preferred way of doing this, or even having an `unsafe_size(size_t)`.

Reply:
@lezcano has provided a great explanation. In numbers:
```
In [5]:  timer = Timer("m1.size(0);", setup="at::Tensor m1=torch::empty({1,1,1});", language="c++", timer=timeit.default_timer)
   ...:  stats=timer.collect_callgrind(number=30, repeats=3)
   ...:  print(stats[1].as_standardized().stats(inclusive=False))
<torch.utils.benchmark.utils.valgrind_wrapper.timer_interface.FunctionCounts object at 0x7f95c87d9400>
  180  torch/include/ATen/core/TensorBody.h:main
  150  ???:c10::TensorImpl::sizes() const
  134  /tmp/benchmark_utils_jit_build__16 ... 814291730550202/timer_src.cpp:main
   60  torch/include/c10/core/WrapDimMinimal.h:main
   60  ???:c10::TensorImpl::dim() const
    4  torch/include/c10/util/irange.h:main

Total: 588

In [6]:  timer = Timer("m1.sizes()[0];", setup="at::Tensor m1=torch::empty({1,1,1});", language="c++", timer=timeit.default_timer)
   ...:  stats=timer.collect_callgrind(number=30, repeats=3)
   ...:  print(stats[1].as_standardized().stats(inclusive=False))
<torch.utils.benchmark.utils.valgrind_wrapper.timer_interface.FunctionCounts object at 0x7f95c885c3a0>
  150  ???:c10::TensorImpl::sizes() const
  105  /tmp/benchmark_utils_jit_build__16 ... 283917667944551/timer_src.cpp:main
   90  torch/include/ATen/core/TensorBody.h:main

Total: 345
```